python/lsst/daf/butler/registry/_butler_registry.py

```diff
@@ -216,3 +219,53 @@ def getDatastoreBridgeManager(self) -> DatastoreRegistryBridgeManager:
             associated datastores.
         """
         raise NotImplementedError()
+
+    @abstractmethod
+    def get_datastore_records(self, ref: DatasetRef) -> DatasetRef:
+        """Retrieve datastore records for given ref.
+
+        Parameters
+        ----------
+        ref : `DatasetRef`
+            Dataset reference for which to retrieve its corresponding datastore
+            records.
+
+        Returns
+        -------
+        updated_ref : `DatasetRef`
+            Dataset reference with filled datastore records.
+
+        Notes
+        -----
+        If this method is called with the dataset ref that is not known to the
+        registry then the reference with an empty set of records is returned.
```

Comment:
I'm a bit surprised by this behavior; generally in methods that add records to refs, I think we want:

- `DatasetRef._datastore_records = None`: this ref might have records, but we haven't tried to "expand" the ref to get them yet;
- `DatasetRef._datastore_Records = {}`: this ref is known to the registry but is not in any datastore and hence has no datastore records;
- (raise an exception): this ref is not known to the registry at all.

But if it's expensive or otherwise inconvenient to maintain that invariant maybe it's not worth it.

Reply:
The complication here is InMemoryDatastore which does not have any records. I want to avoid raising an exception in that case because Butler would need then to figure out what to do with it. This is not ideal of course as we now depend on Butler+Datastore doing correct thing when the records are empty. The alternative is to do an additional check for dataset_id in that case, which is one more query on a large table. I think I want to see additional use cases for this method to decide if this needs change.

Reply:
If the answer is "because InMemoryDatastore" then let's just revisit it when that's dead and gone.